### ISSUE TITLE
🐛 BUG: Global scope is reused for all worker code, different from cloud environment

### ISSUE THREAD

**Issue by aboodman** [cloudflare/workers-sdk#3984]
### Which Cloudflare product(s) does this pertain to?

Wrangler core

### What version(s) of the tool(s) are you using?

3.6.0

### What version of Node are you using?

18.3.0

### What operating system are you using?

Mac

### Describe the Bug

In the production cloudflare environment the execution environment (global scope) for code is sometimes reused, but only across durable objects of the same class.

So for instance if your wrangler.toml declares two Durable Objects with classes `Chess` and `Checkers`, instances of `Checkers` might sometimes be instantiated in a js context that was previously used by a previous instance of `Checkers`, but they will never be instantiated in the js context previously used by an instance of `Chess`.

Further, I'm pretty sure that in the production environment two instances of a DO can never be "running" at the same time inside an execution context (by running here I mean receiving messages). But I have an open question about that here:

https://discord.com/channels/595317990191398933/773219443911819284/1153898195886293122

In Wrangler, however, all code is run in the same js context (global scope) concurrently.

### Please provide a link to a minimal reproduction

https://github.com/rocicorp/cf-do-js-exec-context

**Comment by admah** [comment]
Just to confirm, this is when you're running `wrangler dev`? Do you also see this behavior when running `wrangler dev --remote`?

**Comment by aboodman** [comment]
Yes, the report was from `wrangler dev`. I just checked `wrangler dev --remote` and it has a third, different behavior.

Here is the output from the production cf environment:

```
Worker js exec context id: nlxfszyh
DOClass1 foo response: DO id: 9xuuqefw, js context id: fmz67dwn
DOClass2 bar in your browser of choice. response: DO id: ah16ukh8, js context id: lxw7futo
DOClass1 foo response: DO id: 9xuuqefw, js context id: fmz67dwn
DOClass2 bar in your browser of choice. response: DO id: ah16ukh8, js context id: lxw7futo
```

Note that the worker, DOClass1, and DOClass2 each have their own js context.

Here is the output from `wrangler dev`:

```
Worker js exec context id: qff63q1g
DOClass1 foo response: DO id: pxmht8t8, js context id: qff63q1g
DOClass2 bar in your browser of choice. response: DO id: e23urbkq, js context id: qff63q1g
DOClass1 foo response: DO id: pxmht8t8, js context id: qff63q1g
DOClass2 bar in your browser of choice. response: DO id: e23urbkq, js context id: qff63q1g
```

The worker, DOClass1, and DOClass2 all share the same context.

And here is the output from `wrangler dev --remote`:

```
Worker js exec context id: e3d1rg6q
DOClass1 foo response: DO id: lgwax42, js context id: nrf9h6kh
DOClass2 bar in your browser of choice. response: DO id: 6zti8luc, js context id: nrf9h6kh
DOClass1 foo response: DO id: lgwax42, js context id: nrf9h6kh
DOClass2 bar in your browser of choice. response: DO id: 6zti8luc, js context id: nrf9h6kh
```

The worker has one JS Context, and the two DOs share a JS Context.

**Comment by aboodman** [comment]
You can generate this output yourself running the repo attached to this bug.
